Library/Homebrew/dependency.rb

```diff
@@ -145,3 +145,21 @@ def installed?
     false
   end
 end
+
+class PrioritizedTapDependency < Dependency
+  attr_reader :tap
+
+  def initialize(name, tags=[], env_proc=DEFAULT_ENV_PROC, option_name=name)
+    @tap, _, option_name = option_name.rpartition "/"
+    begin
+      Formulary.factory(name)
+    rescue FormulaUnavailableError
+      @tap = "Homebrew/homebrew"
+    end
```

Comment:
I think it's better to move this logic to `to_formula`. That is only resolving to formula when necessary.

Personally, I think it should be allowed to load formula from other taps(maybe showing a warning in the mean time). Therefore, I think you can take such implementation:

```
class PrioritizedTapDependency < Dependency
  attr_reader :tap

  def initialize(name, tap, tags=[], env_proc=DEFAULT_ENV_PROC, option_name=name)
    @tap = tap
    super name, tags, env_proc, option_name
  end

  def to_formula
    begin
      formula = Formulary.factory("#{tap}/#{name}")
    rescue FormulaUnavailableError
      formula = Formulary.factory(name)
      unless formula.core_formula?
        opoo "<warning message>"
      end
    end
    formula.build = BuildOptions.new(options, formula.options)
    formula
  end
end
```